### PR TITLE
Density test : explicit concurrency parameterization

### PR DESCRIPTION
**Pull request by jayunit100**
Here is a PR for supporting explicit concurrency variations in the E2E density tests #6028 .  This will allow us, by varying just a single parameter, to either test **minion's density capacity** (high pods per minion, w/ lower total pods) as well as test **scheduler's density capacity** (this is done by doing less pods per minion, w/ higher total Pods.

... and of course, existing tests have been ported into the parameterized struct.

Additionally, 
- turns on a simple test of pods == 3 , that way the tests can continue running, and the only thing that needs to be updated is the parameterization.
- removes the magic constants, and instead creates **Density** struct, which defines all the paramaters for various types of density tests.  This has obvious benefits. 

Seems to work for me ... :) !

### CODE REVIEW

**Comment by jayunit100** [comment]
BTW there are a couple of extra commits, I can rebase after another pair of eyes gets a look at this !

File: test/e2e/density.go [hunk 1]
```diff
@@ -210,31 +210,87 @@ var _ = Describe("Density", func() {
...
-		threads := 5
-		var wg sync.WaitGroup
-		wg.Add(threads)
-		for i := 0; i < threads; i++ {
-			go func() {
-				defer wg.Done()
-				for i := 0; i < 10; i++ {
-					name := "my-hostname-thrash-" + string(util.NewUUID())
-					RunRC(c, name, ns, "gcr.io/google_containers/pause:go", 10*minionCount)
+	//This test should always run.
```
Comment (rjnagal) [inline test/e2e/density.go]: nit: run gofmt to fix space in comments.

File: test/e2e/density.go [hunk 2]
```diff
@@ -210,31 +210,87 @@ var _ = Describe("Density", func() {
...
+	//TODO (wojtek-t):don't skip d30 after #6059
+	d30 := Density{totalPods: 30, podsPerMinion: 0, rcsPerThread: 1, skip: true}
+	d50 := Density{totalPods: 50, podsPerMinion: 0, rcsPerThread: 1, skip: true}
+	d100 := Density{totalPods: 100, podsPerMinion: 0, rcsPerThread: 1, skip: true}
+	d500t5 := Density{totalPods: 500, podsPerMinion: 10, rcsPerThread: 5, skip: true}
+	d500t25 := Density{totalPods: 500, podsPerMinion: 10, rcsPerThread: 25, skip: true}
+
+	dtests := []Density{d3, d30, d50, d100, d500t5, d500t25}
+
+	glog.Info("%v %v %v %v %v %v", d3, d30, d50, d100, d500t5, d500t25)
```
Comment (rjnagal) [inline test/e2e/density.go]: nit: Move this in the loop? Printing params before the test starts might be more helpful.

File: test/e2e/density.go [hunk 3]
```diff
@@ -210,31 +210,87 @@ var _ = Describe("Density", func() {
...
+				RCName = "my-hostname-density" + strconv.Itoa(dtest.totalPods) + "-" + string(util.NewUUID())
+				RunRC(c, RCName, ns, "gcr.io/google_containers/pause:go", dtest.totalPods)
+			})
+			glog.Info("moving on, test already finished....")
+		} else {
+			// # of threads calibrate to totalPods
+			threads := (dtest.totalPods / (dtest.podsPerMinion * dtest.rcsPerThread))
+
+			name := fmt.Sprintf(
+				"[Skipped] should be able to launch %v pods, avg %v on each minion, in ",
```
Comment (rjnagal) [inline test/e2e/density.go]: Remove Skipped.

Comment (rjnagal) [inline test/e2e/density.go]: incomplete name? Args mismatch.

**Comment by rjnagal** [comment]
Looks fine. I'll wait in case @wojtek-t wants to take a look. Feel free to squash the commits by then.

**Comment by jayunit100** [comment]
squash+pushed.  thanks for the review !

**Comment by wojtek-t** [comment]
We have a holiday in Poland today - I will take a look tomorrow morning. But I'm fine with merging it - we can apply my comments (if any) in a separate PR.

**Comment by rjnagal** [comment]
@jayunit100 Can you fix the nits I mentioned yesterday? We should be able to merge after that.

**Comment by jayunit100** [comment]
sure !  fixing now

**Comment by jayunit100** [comment]
.... okay, fixed the nits & pushed

**Comment by rjnagal** [comment]
Thanks @jayunit100, merging

File: test/e2e/density.go [hunk 4]
```diff
@@ -210,31 +210,87 @@ var _ = Describe("Density", func() {
...
+		//cannot do a range iterator over structs.
+		dtest := dtests[i]
+
+		glog.Info("Density test parameters: %v", dtest)
+
+		//if ppm==0, its a raw density test.
+		//otherwise, we continue launching n nodes per pod in threads till we meet the totalPods #.
+		if dtest.podsPerMinion == 0 {
+			//basic density tests
+			name := fmt.Sprintf("should allow starting %d pods per node", dtest.totalPods)
```
Comment (wojtek-t) [inline test/e2e/density.go]: I don't think this change is what we would like to do. The purpose of this test is to test how the performance looks in clusters with different sizes. So if we're testing a cluster with 2 nodes and a cluster with 100 nodes, the number of pods that we would like to start should be different (what I mean is that starting 30 pods on 100 nodes cluster doesn't expose Kubelets on heavy load at all).

I will prepare a PR for it.

BTW, I've just discovered that It() with function that takes parameters from "outside" world doesn't work (i.e. there are races). Will try to fix that too.

Comment (wojtek-t) [inline test/e2e/density.go]: Regarding the race, this is the example of a log that shows it does not what we want:

Density 
  should allow starting 3 pods per node
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/density.go:250
[BeforeEach] Density
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/density.go:197
>>> testContext.KubeConfig: /usr/local/google/home/wojtekt/.kube/.kubeconfig
STEP: Making sure all 20 replicas exist
STEP: Making sure all 20 replicas exist
STEP: Making sure all 20 replicas exist
STEP: Making sure all 20 replicas exist
STEP: Making sure all 20 replicas exist
STEP: Making sure all 20 replicas exist
STEP: Making sure all 20 replicas exist
STEP: Making sure all 20 replicas exist
STEP: Making sure all 20 replicas exist
STEP: Making sure all 20 replicas exist
I0407 10:56:31.395932    1059 density.go:110] Controller my-short-lived-podfbc034b9-dd03-11e4-9e2b-a0481cabf39b: Found 0 pods out of 20

**Pull request merged**